project/benchmark/main.cu

```diff
@@ -0,0 +1,242 @@
+
+    std::vector<uint8_t> ciphertext(test_case.plaintext.size());
+
+    return utils::time_it<Timing>([&]
+    {
+        aes::aes_ctr_encrypt(type, 
+                             test_case.key.data(), 
+                             test_case.nonce.data(), 
+                             test_case.plaintext.data(),
+                             test_case.plaintext.size(),
+                             ciphertext.data());
+    });
+}
+
+/**
+ * Measure how long it takes to perform AES-CTR encryption using cuaes.
+ * 
+ * @param test_case The test data to perform encryption on.
+ *
+ * @return The time (in microseconds) the encryption took.
+ */
+Timing time_cuaes(const TestCase& test_case)
+{
+    cuaes::Type type = cuaes::TYPE(test_case.key.size());
+
+    std::vector<uint8_t> host_ciphertext(test_case.plaintext.size());
+
+    return utils::time_it<Timing>([&]
+    {
+        utils::PagedCudaBuffer<uint8_t> device_key(test_case.key.size());
+        utils::PagedCudaBuffer<uint8_t> device_nonce(test_case.nonce.size());
+        utils::PagedCudaBuffer<uint8_t> device_plaintext(test_case.plaintext.size());
+        utils::PagedCudaBuffer<uint8_t> device_ciphertext(device_plaintext.size());
+
+        device_key.copy_from(test_case.key.data());
+        device_nonce.copy_from(test_case.nonce.data());
+        device_plaintext.copy_from(test_case.plaintext.data());
+
+        cuaes::aes_ctr_encrypt(type, 
+                               device_key.ptr(), 
+                               device_nonce.ptr(), 
+                               device_plaintext.ptr(),
+                               device_plaintext.size(),
+                               device_ciphertext.ptr());
+        
+        device_ciphertext.copy_to(host_ciphertext.data());
+    });
+}
+
+//------------------------------------------------------------------------------
+// Main
+//------------------------------------------------------------------------------
+
+int main()
+{
+    try
+    {
+        // Note: Much like the first pancake in a batch, the first CUDA kernel
+        //       is never quite right; it always takes a little bit longer. So
+        //       let's quickly perform a kernel call before we start collecting
+        //       timings so we don't end up with any weird results.
+        TestCase test_case;
+        test_case.key = random_bytes(KEY_SIZE);
+        test_case.nonce = random_bytes(NONCE_SIZE);
+        test_case.plaintext = random_bytes(16);
+        time_cuaes(test_case);
+
+        std::vector<Timing> cpu_timings = {};
+        std::vector<Timing> gpu_timings = {};
+
+        for (auto data_size : DATA_SIZES)
+        {
+            TestCase test_case;
+            test_case.key = random_bytes(KEY_SIZE);
+            test_case.nonce = random_bytes(NONCE_SIZE);
+            test_case.plaintext = random_bytes(data_size.bytes());
+
+            cpu_timings.push_back(time_libaes(test_case));
+            gpu_timings.push_back(time_cuaes(test_case));
```

Comment:
Interesting thought. I just tried a very un-scientific test, and it didn't seem to have any noticeable impact. I wouldn't be surprised if there's some subtle effect, but if there is, I don't think it's significant considering the timing differences between CPU and GPU. deviate by an order of magnitude with larger data sizes.

This would probably be a little easier to spot if my benchmark did multiple runs and gave average runtimes, so it could minimize some of the variance in results.